Необходимо настроить добавление обложки к проекту.

alexFromAC143937 opened the pull request:
Нужно добавить ограничение на размер загружаемого файла (не должен превышать 5 Мб). При загрузке, если разрешение выше при горизонтальном изображении — 1080×566 пикселей, квадратном — 1080×1080 пикселей, то картинка должна ужиматься.

alexFromAC143937 committed:
Необходимо настроить добавление обложки к проекту.
Нужно добавить ограничение на размер загружаемого файла (не должен превышать 5 Мб). При загрузке, если разрешение выше при горизонтальном изображении — 1080×566 пикселей, квадратном — 1080×1080 пикселей, то картинка должна ужиматься.

alexFromAC143937 commented:
![image](https://github.com/user-attachments/assets/b6c2c3df-60f6-488c-b37e-153f0cb6ea5a)

alexFromAC143937 commented:
![image](https://github.com/user-attachments/assets/3ad71680-5f1b-4e99-b7ca-ea5745894ee6)

alexFromAC143937 commented:
![image](https://github.com/user-attachments/assets/8397dd0d-2c06-4121-8c5b-46a926cce108)

alexFromAC143937 commented:
![image](https://github.com/user-attachments/assets/47286789-1143-4259-a63c-8c2e1fc9ffd4)

Alexey-Zimin commented on src/main/java/faang/school/projectservice/service/s3/S3ServiceMock.java:
```diff
@@ -0,0 +1,33 @@
+
+@Configuration
+public class S3ServiceMock {
+
```
Почему в названии присутствует Mock?

alexFromAC143937 replied:
Исправил

Alexey-Zimin commented on src/main/java/faang/school/projectservice/service/ProjectCoverService.java:
```diff
@@ -0,0 +1,84 @@
+        return new ResourceDto(key);
+    }
+
+    public InputStream upload(Long idProject) {
```
стоит одинаково называть входные параметры методов. Здесь у тебя idProject, а чуть выше projectId.

alexFromAC143937 replied:
fixed

Alexey-Zimin commented on src/main/java/faang/school/projectservice/service/ProjectCoverService.java:
```diff
@@ -0,0 +1,84 @@
+    public ResourceDto add(Long projectId, MultipartFile file) {
+        log.info("Start add/update a cover for the project: {}", projectId);
+
+        Project project = projectRepository.getProjectById(projectId);
```
А если проекта по projectId нет, то будет падение.

alexFromAC143937 replied:
Падения не будет, так как запросы по ИД возвращают Optional, и если обработки его нет, то говорит о том, что был вызов уже обработан и будет сгенерирована ошибка, а так как сюда дошло, то говорит, что значени было
public Project getProjectById(Long projectId) {
        return projectJpaRepository.findById(projectId).orElseThrow(
                () -> new EntityNotFoundException(String.format("Project not found by id: %s", projectId))
        );
    }

Alexey-Zimin commented on src/main/java/faang/school/projectservice/service/ProjectCoverService.java:
```diff
@@ -0,0 +1,84 @@
+        log.info("Start upload a cover for the project: {}", idProject);
+
+        Project project = projectRepository.getProjectById(idProject);
+        return s3Service.fromS3File(bucketName, project.getCoverImageId());
```
здесь тоже нет проверки на то, что проекта по запросу может не оказаться

alexFromAC143937 replied:
Здесь она не нужна. Так как результат Optional и уже обработан

Alexey-Zimin commented on src/main/java/faang/school/projectservice/service/ProjectCoverService.java:
```diff
@@ -0,0 +1,84 @@
+        log.info("Start delete a cover for the project: {}", idProject);
+
+        Project project = projectRepository.getProjectById(idProject);
+        if (project.getCoverImageId() != null) {
```
здесь тоже нет проверки на случай отсутствия проекта

alexFromAC143937 replied:
Здесь она не нужна. Так как результат Optional  и уже обработан

Alexey-Zimin commented on src/main/java/faang/school/projectservice/utilities/ImageConvert.java:
```diff
@@ -0,0 +1,51 @@
+
+        if (originalBufferedImage.getHeight() >= originalBufferedImage.getWidth()) {
+            heightNewImage = targetImageSize;
+            widthNewImage = (int) ((double) originalBufferedImage.getWidth()
```
а здесь точно необходимо нужно приводить к double?

alexFromAC143937 replied:
Да. Так как делятс челые числа, то последеления близких чисел будет значение 1 или 0. А нам чем точнее тем лучше, Видится не нужным более 10 символов точность... А в остальном , нужно

Alexey-Zimin commented on src/test/java/faang/school/projectservice/controller/ProjectCoverControllerTest.java:
```diff
@@ -0,0 +1,70 @@
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class ProjectCoverControllerTest {
```
Контроллеры же тестируют с использованием MockMvc. С помощью него дергаются именно ресты, а тебя, как я понял, сейчас дергаются как обычные методы класса.

Alexey-Zimin commented on src/test/java/faang/school/projectservice/service/ProjectCoverServiceTest.java:
```diff
@@ -0,0 +1,139 @@
+
+    @Test
+    void testAddSuccess() throws IOException {
+        int targetImageSize = 100;
```
А неуспешного добавления обложки не может быть? В сервисе у тебя там исключения могут возникнуть, так что, наверное, fail-тест тоже должен быть .

Alexey-Zimin commented on src/test/java/faang/school/projectservice/service/ProjectCoverServiceTest.java:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    @Test
+    void testUploadSuccess() {
```
Тоже, скорее всего, должен быть fail-тест

alexFromAC143937 replied:
В текущем методе не возникает исключение. Тест ошибки выполнен в месте возникновения

alexFromAC143937 committed:
41080 - fix
S3ServiceImplTest, S3ServiceConnectionConfiguration, ProjectCoverServiceTest, ProjectCoverService, ProjectCoverControllerTest

jibulani reviewed (dismissed):
хорошее покрытие тестами, молодец!

jibulani commented:
осталось поправить конфликт

alexFromAC143937 committed:
Merge branch 'Murloc-master-stream7' into Murloc7-BAlex-uploadFile-BJS2-41080
